src/shared/functions.ts

```diff
@@ -0,0 +1,298 @@
+    } catch (err) {
+      throw new Error(
+        'Could not convert legacy request. Please switch to a general signing ' +
+        'request. See gridplus-sdk docs for more information.',
+      );
+    }
+    data = {
+      fwConstants,
+      encodingType: EXTERNAL.SIGNING.ENCODINGS.EVM,
+      curveType: EXTERNAL.SIGNING.CURVES.SECP256K1,
+      hashType: EXTERNAL.SIGNING.HASHES.KECCAK256,
+      signerPath: data.signerPath,
+      payload,
+    };
+    return {
+      request: buildGenericSigningMsgRequest({ ...data, fwConstants }),
+      isGeneric: true,
+    };
+  } else if (currency === 'ETH') {
+    // Legacy signing pathway -- should deprecate in the future
+    return {
+      request: ethereum.buildEthereumTxRequest({ ...data, fwConstants }),
+      isGeneric: false,
+    };
+  } else if (currency === 'ETH_MSG') {
+    return {
+      request: ethereum.buildEthereumMsgRequest({ ...data, fwConstants }),
+      isGeneric: false,
+    };
+  } else if (currency === 'BTC') {
+    return {
+      request: bitcoin.buildBitcoinTxRequest({ ...data, fwConstants }),
+      isGeneric: false,
+    };
+  }
+  return {
+    request: buildGenericSigningMsgRequest({ ...data, fwConstants }),
+    isGeneric: true,
+  };
+};
+
+export const request = async ({
+  url,
+  payload,
+  timeout = 60000,
+}: RequestParams) => {
+  return superagent
+    .post(url)
+    .timeout(timeout)
+    .send({ data: payload })
+    .catch(validateRequestError)
+    .then(async (response) => {
+      // Handle formatting or generic HTTP errors
+      if (!response.body || !response.body.message) {
+        throw new Error(`Invalid response:  ${response.body.message}`);
+      } else if (response.body.status !== 200) {
+        throw new Error(
+          `Error code ${response.body.status}: ${response.body.message}`,
+        );
+      }
+
+      const { data, errorMessage, responseCode } = parseLattice1Response(
+        response.body.message,
+      );
+
+      if (errorMessage || responseCode) {
+        throw new LatticeResponseError(responseCode, errorMessage);
+      }
+
+      return data;
+    })
+};
+
+/**
+ * `sleep()` returns a Promise that resolves after a given number of milliseconds.
+ */
+function sleep (ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/**
+ * `retryWrapper()` retries a function call if the error message or response code is present and the
+ * number of retries is greater than 0.
+ *
+ * @param fn - The function to retry
+ * @param params - The parameters to pass to the function
+ * @param retries - The number of times to retry the function
+ * @param client - The Client to use for side-effects
+ */
+export const retryWrapper = async ({
+  fn,
+  params,
+  retries,
+  client,
+}) => {
+  return fn({ ...params }).catch(async (err) => {
+    const errorMessage = err.errorMessage;
+    const responseCode = err.responseCode;
+
+    if ((errorMessage || responseCode) && retries) {
```

Comment:
Is `responseCode == 0` in normal circumstances? Where does that come from?

Reply:
`responseCode` is sent from the backend services. For a successful request, it is not included in the response, so it is usually `undefined` unless an error happens.

Reply:
A comment would be good in case we accidentally change this in the backend and get confused about what's happening